### PR TITLE
Adds vmservices for getting iOS build options

### PR DESCRIPTION
**Pull request by chunhtai**
part of https://github.com/flutter/flutter/issues/120405

The goal is to provide ios build options for devtool, so that it can display these options for the developers to choose from.

There will be a new vmservices to get app settings given a build option.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter_tools/lib/executable.dart [hunk 1]
```diff
@@ -33,6 +33,7 @@ import 'src/commands/generate.dart';
 import 'src/commands/generate_localizations.dart';
 import 'src/commands/ide_config.dart';
 import 'src/commands/install.dart';
+import 'src/commands/list_xcode_modes.dart';
```
Comment (jmagman) [inline packages/flutter_tools/lib/executable.dart:36]: This file is missing.

**Comment by christopherfujino** [comment]
As a high level question, would it make sense for this message to fetch info on android too (if configured)? or is that not needed?

**Comment by chunhtai** [comment]
> As a high level question, would it make sense for this message to fetch info on android too (if configured)? or is that not needed?

There will be a different sets of vm services for Android. I decided to separate them out because the response format will be different. Plus Android and iOS will have different UI because the validation checks are different.

We could group them like
```json
{
  result: success,
  android: {...},
  iOS: {...},
}
```

however I can't find a good reason to do that, the caller might as well call two different vm serivces to get the result. It is also likely we will display android and iOS in separate tab, it may only need result from one platform.

We could reuse this for macOS though since its format is almost the same as iOS, but I don't think we should do it unless we decide to reuse the same tab to display iOS and macOS result.

**Comment by christopherfujino** [comment]
> > As a high level question, would it make sense for this message to fetch info on android too (if configured)? or is that not needed?
> 
> There will be a different sets of vm services for Android. I decided to separate them out because the response format will be different. Plus Android and iOS will have different UI because the validation checks are different.
> 
> We could group them like
> 
> ```json
> {
>   result: success,
>   android: {...},
>   iOS: {...},
> }
> ```
> 
> however I can't find a good reason to do that, the caller might as well call two different vm serivces to get the result. It is also likely we will display android and iOS in separate tab, it may only need result from one platform.
> 
> We could reuse this for macOS though since its format is almost the same as iOS, but I don't think we should do it unless we decide to reuse the same tab to display iOS and macOS result.

fair enough. you've convinced me :)

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 2]
```diff
@@ -943,6 +946,14 @@ abstract class ResidentHandlers {
     return sharedSkSlWriter(device, data);
   }
 
+  /// Write the SkSL shaders to a zip file in build directory.
+  ///
+  /// Returns the name of the file, or `null` on failures.
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/resident_runner.dart:951]: I'm guessing this is the wrong dartdoc? :)

**Review by christopherfujino — APPROVED** [review]
LGTM

**Review by christopherfujino — CHANGES_REQUESTED** [review]
code lgtm, but please update the dartdoc

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 3]
```diff
@@ -181,6 +186,7 @@ Future<vm_service.VmService> setUpVmService(
   CompileExpression? compileExpression,
   Device? device,
   GetSkSLMethod? skSLMethod,
+  GetXcodeProjectInfo? getIOSProjectInfo,
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/vmservice.dart:189]: This method signature is getting unwieldy `... null, null, device!.device, null, null...`
Can you swap these all to named parameters?

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 4]
```diff
@@ -283,6 +289,30 @@ Future<vm_service.VmService> setUpVmService(
...
+        };
+      }
+      return <String, Object>{
+        'result': <String, Object>{
+          'type': 'Success',
+          'targets': info.targets,
+          'schemes': info.schemes,
+          'buildConfigurations': info.buildConfigurations,
+        },
+      };
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/vmservice.dart:310]: This seems untested?

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 5]
```diff
@@ -283,6 +289,30 @@ Future<vm_service.VmService> setUpVmService(
...
     registrationRequests.add(vmService.registerService('flutterGetSkSL', 'Flutter Tools'));
   }
+
+  if (getIOSProjectInfo != null) {
+    vmService.registerServiceCallback('flutterGetIOSBuildOptions', (Map<String, Object?> params) async {
+      final XcodeProjectInfo? info = await getIOSProjectInfo();
+      if (info == null) {
+        return <String, Object>{
+          'result': <String, Object>{
+            'type': 'Success',
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/vmservice.dart:299]: Should it be a success if the iOS project details can't be determined?

Comment (chunhtai) [reply packages/flutter_tools/lib/src/vmservice.dart]: This success is a type of response, that means everything complete without error.
See
https://github.com/dart-lang/sdk/blob/a22e003ee1b9fe5d921536d9401829e9ebdefc2a/pkg/vm_service/lib/src/vm_service.dart#L113

**Comment by chunhtai** [comment]
Instead of drill through the callback, I decided to drill through the FlutterProject given how many layers I need to drill through, and the number of vmservices I will have to add in the future.

**Comment by chunhtai** [comment]
A friendly bump 👍

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 6]
```diff
@@ -249,6 +249,7 @@ class FlutterDevice {
     Restart? restart,
     CompileExpression? compileExpression,
     GetSkSLMethod? getSkSLMethod,
+    FlutterProject? flutterProject,
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/resident_runner.dart:252]: does this argument ever get referenced?

Comment (chunhtai) [reply packages/flutter_tools/lib/src/resident_runner.dart]: nice catch, I confused myself with all the drilling. fixed

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 7]
```diff
@@ -270,19 +277,43 @@ Future<vm_service.VmService> setUpVmService(
...
+        'result': <String, Object>{
+          kResultType: kResultTypeSuccess,
+          'targets': info.targets,
+          'schemes': info.schemes,
+          'buildConfigurations': info.buildConfigurations,
+        },
+      };
+    });
+    registrationRequests.add(
+        vmService.registerService('flutterGetIOSBuildOptions', 'Flutter Tools'));
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/vmservice.dart:314]:
```suggestion
      vmService.registerService('flutterGetIOSBuildOptions', 'Flutter Tools'),
    );
```

**Review by christopherfujino — APPROVED** [review]
nit about formatting, otherwise LGTM

**Review by jmagman — APPROVED** [review]
Thanks for swapping to named parameters, much easier to read 🙂

**Pull request merged**